crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -797,6 +826,55 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.types.definitions.insert(definition, context_expr_ty);
     }
 
+    /// Infer the types of exceptions caught by an exception handler
+    fn infer_handled_exception_types(
+        &self,
+        handled_exceptions: &'db ast::Expr,
+    ) -> impl Iterator<Item = ExceptHandlerType<'db>> + '_ {
+        let exception_nodes = match handled_exceptions {
+            ast::Expr::Tuple(multiple_exceptions) => multiple_exceptions.elts.as_slice(),
+            single_exception => std::slice::from_ref(single_exception),
+        };
+
+        exception_nodes.iter().map(|node| {
+            let node_ty = self
+                .types
+                .expression_ty(node.scoped_ast_id(self.db, self.scope));
+
+            match node_ty {
+                Type::Any | Type::Unknown => ExceptHandlerType::Valid {
+                    exception_ty: node_ty,
+                },
+                // TODO: we should check whether `exception_class`
+                // is actually a subclass of `builtins.BaseException` --Alex
+                Type::Class(class_ty) => ExceptHandlerType::Valid {
+                    exception_ty: Type::Instance(class_ty),
+                },
+                // TODO: unions should also be valid --Alex
+                _ => ExceptHandlerType::Invalid { node, node_ty },
```

Comment:
Yeah, I realised after sleeping on it that my current approach of distinguishing between a literal tuple in the AST and other nodes is hopelessly naive. You can of course have dynamic tuples in `except` blocks:

```pycon
>>> EXCEPTIONS = (TypeError, AttributeError)
>>> try:
...     object.foo
... except EXCEPTIONS:
...     pass
... 
>>> 
```

So it may be better to just axe all this type-checking until we have the tools to do it properly. There's certainly no way we can check if something is a consistent subtype of `tuple[type[BaseException], ...] | type[BaseException]` right now, and only permitting a literal tuple of exceptions else risks emitting false positives on user code :(